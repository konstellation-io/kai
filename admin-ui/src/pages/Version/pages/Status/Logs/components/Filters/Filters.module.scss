@import '../../../../../../../styles/colors';
@import '../../../../../../../styles/mixins';
@import '../../../../../../../styles/variables';

.container {
  @include border-header;
  @include font-body;
  display: flex;
  align-items: center;
  justify-content: space-between;
  height: $grid-unit * 6;
  flex-shrink: 0;
}

.filter {
  @include font-code;
  display: flex;
  flex-direction: row;
  padding: 0 16px;

  background-color: $font-color-dark;
  color: $font-color-white;

  div:first-child {
    margin-right: 8px;
  }
}

.leftFilters {
  display: flex;
  align-items: center;
  margin-left: $grid-unit * 2.5;
}
.rightFilters {
  margin-right: $grid-unit * 2.5;
  display: flex;
  & > div:not(:last-child) {
    border-right: 2px solid #1d1e23;
    border-right-style: inset;
  }
}
.dateFilter {
  width: $grid-unit * 21.5;
}
