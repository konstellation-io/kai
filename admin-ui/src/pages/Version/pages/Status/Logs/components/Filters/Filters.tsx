import React from 'react';
import styles from './Filters.module.scss';
import DateFilter from './components/DatesFilter/DateFilter';
import { TabFilters } from '../../../../../../../graphql/client/queries/getLogs.graphql';
import Button from '../../../../../../../components/Button/Button';
import Right from '../../../../../../../components/Layout/Right/Right';

type Props = {
  onDateChange: Function;
  filterValues: TabFilters;
};
function Filters({ onDateChange, filterValues }: Props) {
  return (
    <div className={styles.container}>
      <Right style={styles.rightFilters}>
        <DateFilter
          selectedOption={filterValues.dateOption}
          onDateChange={onDateChange}
        />
        <Button label="JUMP TO NOW" />
      </Right>
    </div>
  );
}

export default Filters;
