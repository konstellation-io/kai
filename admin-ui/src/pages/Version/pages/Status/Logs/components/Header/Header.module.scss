@import '../../../../../../../styles/mixins';
@import '../../../../../../../styles/funcs';
@import '../../../../../../../styles/colors';

.container {
  display: flex;
  align-items: center;

  transition: height ease 0.3s;

  &.opened {
    height: 50px;
    flex-shrink: 0;

    .buttons {
      opacity: 1;
    }
  }
}

.title {
  @include font-caption;
  color: $font-color-light;

  display: flex;
  align-items: center;
  margin: 0 20px 0 20px;

  span {
    font-size: 14px;
  }

  svg {
    margin-right: 8px;
  }
}

.runtime {
  @include font-body;
  color: $font-color-white;
}

.version {
  @include font-body;
  color: $font-color-light;

  margin-left: 16px;
}

.buttons {
  display: flex;
  flex-direction: row;
  transition: opacity ease 0.4s;
  margin: 4px 8px 0 auto;

  div {
    margin-left: 8px;
    cursor: pointer;

    &:hover {
      color: $font-color-white;
    }
  }
}

.stickBottom {
  transition: color ease 0.4s, background-color ease 0.4s;

  &.active {
    color: $font-color-white;
    background-color: $bg-color-lighter;

    &:hover {
      color: $font-color-light;
    }
  }
}
