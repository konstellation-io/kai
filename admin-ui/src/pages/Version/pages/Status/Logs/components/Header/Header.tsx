import React from 'react';

import IconOpen from '@material-ui/icons/ExpandLess';
import IconClose from '@material-ui/icons/ExpandMore';
import IconStickBottom from '@material-ui/icons/VerticalAlignBottom';
import IconClear from '@material-ui/icons/DeleteOutline';
import IconLogs from '@material-ui/icons/ListAlt';

import { useApolloClient } from '@apollo/react-hooks';

import cx from 'classnames';
import styles from './Header.module.scss';

type Props = {
  togglePanel: () => void;
  opened: boolean;
  stickToBottom: boolean;
  toggleStickToBottom: () => void;
};
function Header({
  togglePanel,
  opened,
  stickToBottom,
  toggleStickToBottom
}: Props) {
  const client = useApolloClient();

  function clearLogs(): void {
    client.writeData({ data: { logs: [] } });
  }

  const Icon = opened ? IconClose : IconOpen;

  return (
    <div
      className={cx(styles.container, {
        [styles.opened]: opened
      })}
    >
      <div className={styles.title}>
        <IconLogs className="icon-regular" />
        <span>Logs console</span>
      </div>
      <div className={styles.buttons}>
        <div onClick={clearLogs}>
          <IconClear className="icon-regular" />
        </div>
        <div
          className={cx(styles.stickBottom, { [styles.active]: stickToBottom })}
          onClick={toggleStickToBottom}
        >
          <IconStickBottom className="icon-regular" />
        </div>
        <div onClick={togglePanel}>
          <Icon className="icon-regular" />
        </div>
      </div>
    </div>
  );
}

export default Header;
