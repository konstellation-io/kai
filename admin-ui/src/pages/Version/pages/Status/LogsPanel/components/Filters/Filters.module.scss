@import '../../../../../../../styles/colors';
@import '../../../../../../../styles/mixins';
@import '../../../../../../../styles/variables';

.container {
  @include font-body;
  line-height: 0;

  display: flex;
  flex-wrap: wrap;
  align-items: center;
  justify-content: space-between;
  height: auto;
  padding: 0 20px;
  flex-shrink: 0;
}

.filterContainer {
  display: flex;
  flex-direction: row;
  align-items: center;

  > div {
    margin-right: $grid-unit;
  }
}
.rightFilters {
  flex-shrink: 0;

  .filterLevel {
    width: 142px;
  }

  .dateFilter {
    width: $grid-unit * 21.5;
  }
}

.leftContainer {
  flex-grow: 1;
  flex-basis: 0;
  max-width: 100 * $grid-unit;

  .searchFilter {
    width: 60%;
    height: 30px;
    min-width: 100px;
    margin: $grid-unit $grid-unit $grid-unit 0;

    .searchForm {
      input {
        height: 30px;
        min-height: 30px;
      }
    }
  }

  .selectProcesses {
    width: 40%;
    min-width: 170px;
  }
}
