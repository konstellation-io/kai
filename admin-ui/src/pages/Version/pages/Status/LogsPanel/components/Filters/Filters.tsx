import React from 'react';
import styles from './Filters.module.scss';
import DateFilter from './components/DatesFilter/DateFilter';
import Right from '../../../../../../../components/Layout/Right/Right';
import Left from '../../../../../../../components/Layout/Left/Left';
import SearchSelect from '../../../../../../../components/Form/SearchSelect/SearchSelect';
import MultiSelect, {
  MultiSelectOption,
  SelectTheme
} from '../../../../../../../components/Form/MultiSelect/MultiSelect';
import GroupSelect, {
  GroupSelectData
} from '../../../../../../../components/Form/GroupSelect/GroupSelect';
import cx from 'classnames';
import { LogLevel } from '../../../../../../../graphql/types/globalTypes';
import useWorkflowsAndNodes from '../../../../../../../hooks/useWorkflowsAndNodes';
import { NodeSelection } from '../../../../../../../graphql/client/typeDefs';
import { GetLogTabs_logTabs_filters } from '../../../../../../../graphql/client/queries/getLogs.graphql';
import LevelIcon from '../../../../../../../components/LevelIcon/LevelIcon';

// TODO: use GroupSelectData int he filters instead of [string]
function nodesSelectionToDoubleSelector(
  selections: NodeSelection[]
): GroupSelectData {
  const data: { [key: string]: string[] } = {};

  selections.forEach(({ workflowName, nodeNames }) => {
    data[workflowName] = nodeNames;
  });

  return data;
}

function doubleSelectorToNodesSelection(
  selections: GroupSelectData
): NodeSelection[] {
  const data: NodeSelection[] = Object.entries(selections).map(
    ([workflowName, nodeNames]) => ({
      workflowName,
      nodeNames,
      __typename: 'NodeSelection'
    })
  );

  return data;
}

const levelsOrdered = [
  LogLevel.INFO,
  LogLevel.DEBUG,
  LogLevel.WARN,
  LogLevel.ERROR
];

type Props = {
  versionId: string;
  updateFilters: Function;
  filterValues: GetLogTabs_logTabs_filters;
};
function Filters({ updateFilters, filterValues, versionId }: Props) {
  const { workflowsAndNodesNames } = useWorkflowsAndNodes(versionId);

  const levelOptions = levelsOrdered.map(
    (level: LogLevel) =>
      ({
        label: level,
        Icon: <LevelIcon level={level} />
      } as MultiSelectOption<LogLevel>)
  );

  function onNodeSelection(newSelection: GroupSelectData) {
    updateFilters({
      nodes: doubleSelectorToNodesSelection(newSelection)
    });
  }

  function onLevelSelection(newLevels: LogLevel[]) {
    updateFilters({ levels: newLevels });
  }

  function onSearchUpdate(newSearch: string) {
    updateFilters({ search: newSearch });
  }

  return (
    <div className={styles.container}>
      <Left className={cx(styles.filterContainer, styles.leftContainer)}>
        <div className={styles.searchFilter}>
          <SearchSelect
            options={[]}
            onChange={onSearchUpdate}
            value={filterValues.search}
            className={styles.searchForm}
            placeholder="Search"
            hideError
            hideLabel
            showSearchIcon
          />
        </div>
        <div className={styles.selectProcesses}>
          <GroupSelect
            options={workflowsAndNodesNames}
            formSelectedOptions={nodesSelectionToDoubleSelector(
              filterValues.nodes || []
            )}
            onChange={onNodeSelection}
            label=""
            placeholder="Select Processes"
            hideError
            hideSelections
          />
        </div>
      </Left>
      <Right className={cx(styles.filterContainer, styles.rightFilters)}>
        <div className={styles.filterLevel}>
          <MultiSelect<LogLevel>
            onChange={onLevelSelection}
            label=""
            hideError
            placeholder="ALL LEVELS"
            selectionUnit="LEVEL"
            selectAllText="ALL LEVELS"
            options={levelOptions}
            formSelectedOptions={filterValues.levels || []}
            className={styles.selectTypeForm}
            theme={SelectTheme.LIGHT}
          />
        </div>
        <DateFilter
          selectedOption={filterValues.dateOption}
          updateFilters={updateFilters}
          formStartDate={filterValues.startDate}
          formEndDate={filterValues.endDate}
        />
      </Right>
    </div>
  );
}

export default Filters;
