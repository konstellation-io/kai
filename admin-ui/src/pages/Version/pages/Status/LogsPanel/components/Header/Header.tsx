import React from 'react';

import IconOpen from '@material-ui/icons/ExpandLess';
import IconClose from '@material-ui/icons/ExpandMore';
import * as ICONS from '../../../../../../../constants/icons';

import cx from 'classnames';
import styles from './Header.module.scss';
import SvgIcon from '@material-ui/core/SvgIcon';

type Props = {
  togglePanel: () => void;
  opened: boolean;
};
function Header({ togglePanel, opened }: Props) {
  const Icon = opened ? IconClose : IconOpen;

  return (
    <div
      className={cx(styles.container, {
        [styles.opened]: opened
      })}
    >
      <div className={styles.title}>
        <SvgIcon className="icon-small">
          <path d={ICONS.TERMINAL} />
        </SvgIcon>
        <span>LOGS CONSOLE</span>
      </div>
      <div className={styles.buttons}>
        <div onClick={togglePanel}>
          <Icon className="icon-regular" />
        </div>
      </div>
    </div>
  );
}

export default Header;
