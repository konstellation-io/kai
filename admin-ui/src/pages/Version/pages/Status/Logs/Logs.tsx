import React, { useState, useEffect } from 'react';

import Header from './components/Header/Header';
import Filters from './components/Filters/Filters';
import LogsList from './components/LogsList/LogsList';

import cx from 'classnames';
import styles from './Logs.module.scss';
import { useQuery } from '@apollo/react-hooks';
import { GET_CURRENT_LOG_PANEL } from '../../../../../graphql/client/queries/getCurrentLogPanel';

function Logs() {
  const [opened, setOpened] = useState<boolean>(false);
  const [stickToBottom, setStickToBottom] = useState<boolean>(false);
  const { data } = useQuery(GET_CURRENT_LOG_PANEL);

  useEffect(() => {
    setOpened(data && data.logPanel && data.logPanel.nodeId !== undefined);
  }, [data]);

  function togglePanel() {
    setOpened(!opened);
  }

  function scrollToBottom() {
    const listContainer = document.getElementById('VersionLogsListContainer');
    if (listContainer) {
      listContainer.scrollTo({
        top: listContainer.scrollHeight,
        behavior: 'smooth'
      });
    }
  }

  function toggleStickToBottom() {
    if (!stickToBottom) {
      scrollToBottom();
    }

    setStickToBottom(!stickToBottom);
  }

  function onLogsUpdate() {
    if (stickToBottom) {
      scrollToBottom();
    }
  }
  const logPanel = data && data.logPanel;
  const hidden = !logPanel;
  return (
    <>
      <div
        className={cx(styles.container, {
          [styles.opened]: opened,
          [styles.hidden]: hidden
        })}
      >
        <Header
          togglePanel={togglePanel}
          opened={opened}
          stickToBottom={stickToBottom}
          toggleStickToBottom={toggleStickToBottom}
        />
        <div className={cx(styles.content, { [styles.opened]: opened })}>
          {logPanel && (
            <>
              <Filters filters={{ node: logPanel.nodeName }} />
              <LogsList
                nodeId={logPanel.nodeId}
                runtimeId={logPanel.runtimeId}
                onUpdate={onLogsUpdate}
              />
            </>
          )}
        </div>
      </div>
      <div
        className={cx(styles.shield, { [styles.show]: opened })}
        onClick={togglePanel}
      />
    </>
  );
}

export default Logs;
