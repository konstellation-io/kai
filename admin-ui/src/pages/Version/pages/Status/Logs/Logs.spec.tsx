import React from 'react';
import Logs from './Logs';
import { shallow } from 'enzyme';
import Filters from './components/Filters/Filters';
import Header from './components/Header/Header';
import { useQuery } from '@apollo/react-hooks';

const mockLogsPanel = {
  data: {
    logPanel: {
      runtimeId: 'runtimeIdMock',
      nodeId: 'nodeIdMock',
      nodeName: 'nodeNameMock'
    }
  }
};

jest.mock('@apollo/react-hooks', () => ({
  useQuery: jest.fn(() => mockLogsPanel),
  useApolloClient: jest.fn(() => ({ writeData: jest.fn() }))
}));

describe('Logs', () => {
  let wrapper: any;
  const mockSetSelectedNode = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<Logs />);
  });

  it('matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('show right components', () => {
    expect(wrapper.exists('.container')).toBeTruthy();
    expect(wrapper.exists(Header)).toBeTruthy();
    expect(wrapper.exists(Filters)).toBeTruthy();
  });

  it('should be empty when there is no node', () => {
    // Arrange.
    useQuery.mockReturnValueOnce({ data: undefined });

    // Act.
    const wrapper2 = shallow(<Logs />);

    // Assert.
    expect(useQuery).toBeCalled();
    expect(wrapper2).toMatchSnapshot();
  });

  // FIXME: fix it when you can mock or use useEffect with shallow
  it.skip('calls setSelectedNode on shield click', () => {
    expect(wrapper.exists(Filters)).toBeTruthy();

    wrapper.find('.shield').simulate('click');

    expect(mockSetSelectedNode).toHaveBeenCalledTimes(1);
    expect(mockSetSelectedNode).toHaveBeenCalledWith(undefined);
  });

  it('toggleStickToBottom', () => {
    expect(wrapper.find(Header).prop('stickToBottom')).toBeFalsy();

    wrapper.find(Header).prop('toggleStickToBottom')();

    expect(wrapper.find(Header).prop('stickToBottom')).toBeTruthy();
  });
});
