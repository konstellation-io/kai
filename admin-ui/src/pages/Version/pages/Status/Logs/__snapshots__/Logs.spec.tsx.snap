// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logs matches snapshot 1`] = `
<Fragment>
  <div
    className="container"
  >
    <Header
      onClearClick={[Function]}
      opened={false}
      togglePanel={[Function]}
    />
    <div
      className="content"
    >
      <Filters
        filters={
          Object {
            "node": "nodeNameMock",
          }
        }
      />
      <LogsList
        nodeId="nodeIdMock"
        runtimeId="runtimeIdMock"
      />
    </div>
  </div>
  <div
    className="shield"
    onClick={[Function]}
  />
</Fragment>
`;

exports[`Logs should be empty when there is no node 1`] = `
<Fragment>
  <div
    className="container hidden"
  >
    <Header
      onClearClick={[Function]}
      opened={false}
      togglePanel={[Function]}
    />
    <div
      className="content"
    />
  </div>
  <div
    className="shield"
    onClick={[Function]}
  />
</Fragment>
`;
